posthog/api/query.py

```diff
@@ -183,3 +192,121 @@ def _tag_client_query_id(self, query_id: str | None):
 
         tag_queries(client_query_id=query_id)
         set_tag("client_query_id", query_id)
+
+
+ASYNC_FALLBACK_TO_POLLING_TIMEOUT = 50
+
+
+async def query_async(request: Request, *args, **kwargs) -> HttpResponse:
+    """Async endpoint for handling event source queries."""
+
+    try:
+        data = json.loads(request.body)
+    except json.JSONDecodeError:
+        return HttpResponse(
+            json.dumps({"error": "Invalid JSON in request body"}),
+            content_type="application/json",
+            status=400,
+        )
+
+    try:
+        data = QueryRequest.model_validate(data)
+    except ValidationError as exc:
+        capture_exception(exc)
+        return HttpResponse(
+            json.dumps({"error": "JSON parse error - {}".format(str(exc))}), content_type="application/json", status=400
+        )
+
+    check_auth_view = await sync_to_async(QueryViewSet.as_view)({"post": "check_auth_for_async"}, **kwargs)
+    check_auth = await sync_to_async(check_auth_view)(request)
+    if check_auth.status_code != 200 or request.user.is_anonymous:
+        return HttpResponse(check_auth.rendered_content, status=check_auth.status_code)
+
+    team_id = kwargs.get("team_id")
+
+    team = await Team.objects.aget(id=team_id)
+
+    if data.filters_override is not None:
+        data.query = apply_dashboard_filters_to_dict(data.query.model_dump(), data.filters_override.model_dump(), team)
+
+    if data.variables_override is not None:
+        if isinstance(data.query, BaseModel):
+            query_as_dict = data.query.model_dump()
+        else:
+            query_as_dict = data.query
+
+        data.query = apply_dashboard_variables_to_dict(query_as_dict, data.variables_override, team)
+
+    client_query_id = data.client_query_id or uuid.uuid4().hex
+    execution_mode = execution_mode_from_refresh(data.refresh)
+
+    tag_queries(client_query_id=client_query_id)
+    set_tag("client_query_id", client_query_id)
+
+    if execution_mode == execution_mode.CACHE_ONLY_NEVER_CALCULATE:
+        execution_mode = ExecutionMode.RECENT_CACHE_CALCULATE_BLOCKING_IF_STALE
+
+    tag_queries(query=data.query.model_dump())
```

Comment:
GitHub casually ate my comment: 🙃

So going the standalone async function, I think we can simplify things by reusing `QueryViewSet`'s create as it is – i.e. refactoring this handler to do just two things:

1. Pass the request to `QueryViewSet` wraped in `sync_to_async()`, which will run it through checks and `create`
2. If the response from 1 is an in-progress query status, then poll async for {55s - however_long_step_1_took} to return ASAP – if it's anything other than an in-progress query status or polling has timed out, return immediately with what we've got

This way updating the `query` endpoint (e.g. by product analytics folks) won't be a nightmare while we're trying this out.

What do you think?

Reply:
yeah great point, have changed it!